daemon/cmd/ipam.go

```diff
@@ -379,7 +383,24 @@ func (d *Daemon) allocateHealthIPs() error {
 			node.SetEndpointHealthIPv6(result.IP)
 		}
 
+		// Coalescing multiple CIDRs. GH #18868
+		if option.Config.EnableIPv6Masquerade &&
+			option.Config.IPAM == ipamOption.IPAMENI &&
+			result != nil &&
+			len(result.CIDRs) > 0 {
+			result.CIDRs = coalesceCIDRs(result.CIDRs)
+		}
+
 		log.Debugf("IPv6 health endpoint address: %s", result.IP)
+
+		// In ENI and AlibabaCloud ENI mode, we require the gateway, CIDRs, and the ENI MAC addr
+		// in order to set up rules and routes on the local node to direct
+		// endpoint traffic out of the ENIs.
+		if option.Config.IPAM == ipamOption.IPAMENI || option.Config.IPAM == ipamOption.IPAMAlibabaCloud {
+			if d.healthEndpointRouting, err = parseRoutingInfo(result); err != nil {
+				log.WithError(err).Warn("Unable to allocate health information for ENI")
+			}
+		}
```

Comment:
Given the amount of duplication here, have you considered refactoring out the body of `allocateHealthIPs` into a family-agnostic function similar to `allocateDatapathIPs`?